# Asg-10 focuses on questions graded by the mentioned TAs

 

In [41]:
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
import sys

# -- Detect if in Own Install or in Colab
try:
    import google.colab
    OWN_INSTALL = False
except:
    OWN_INSTALL = True
    
if OWN_INSTALL:
    print("Own install")
    sys.path[0:0] = ['../../../../..',  '../../../../../3rdparty',
                     '../../../..',  '../../../../3rdparty',
                     '../../..',     '../../../3rdparty',
                     '../..',        '../../3rdparty',
                     '..',           '../3rdparty' ]
    
else: # In colab
    ! if [ ! -d Jove ]; then git clone https://github.com/ganeshutah/Jove Jove; fi
    sys.path.append('./Jove')
    sys.path.append('./Jove/jove')

# -- common imports --
from jove.PcpJupyterNew import *
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Own install


# =============

In [42]:
help(pcp_solve)

Help on function pcp_solve in module jove.PcpJupyterNew:

pcp_solve(pcp_pairs, OWN_INSTALL=False, run=None, ni=False, di=None, depth=None, tiles_per_row=15)
    Forward user input to a file, which we then use Ling Zhao's pcp solver to solve.
    :param OWN_INSTALL: if True, running on one's own machine; default is Colab (so false)
    :param pcp_pairs: List of tuple pairs representing pcp 'tiles'
    :param run: Number of runs to perform.
    :param ni: No iterative search.
    :param di: Depth increment.
    :param depth: Search depth.
    :param tiles_per_row: Number of tiles to show in single row together (useful for creating more meaningful output)



# (1) LT, 25%:
 

## 1(a): 

In [43]:
# Tool Solution
pcp_solve([('110', '1'), ('1', '0'), ('0', '110')], OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[3, 2, 2, 3, 3, 1, 3, 2, 2, 2, 1, 3, 3, 3, 2, 1, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 1, 3, 2, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 2, 3, 3, 2, 1, 3, 2, 1, 1, 3, 3, 1, 1, 3, 2, 1, 1, 1, 1, 2, 2, 1, 1, 2, 2, 3, 1, 3, 3, 1, 2, 1, 1, 2, 1, 1]
0   1   1   0   0   011 0   1   1   1   011 0   0   0   1   
011 0   0   011 011 1   011 0   0   0   1   011 011 011 0   

011 0   1   1   011 0   1   011 0   0   1   011 0   1   0   
1   011 0   0   1   011 0   1   011 011 0   1   011 0   011 

1   1   011 0   1   011 0   0   1   1   0   0   1   011 0   
0   0   1   011 0   1   011 011 0   0   011 011 0   1   011 

1   011 011 0   0   011 011 0   1   011 011 011 011 1   1   
0   1   1   011 011 1   1   011 0   1   1   1   1   0   0   

011 011 

Solution length: 75 for both solutions

## 1(b): 

In [44]:
# Tool solution
pcp_solve([('110', '1'), ('1', '0'), ('0', '110'), ('0', '1')],
          OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[3, 2, 2, 4, 1]
0   1   1   0   011 
011 0   0   1   1   




Solution length: 5

## 1(c, hand): 

<!-- Solution by hand is 2, 1, 3 -->

We have the following tiles:

  * Tile 1: `(1, 0)`

  * Tile 2: `(101, 1)`

  * Tile 3: `(0, 110)`

Solution construction:

  * Tile 2 is the only tile where both the top and the bottom start with the same character ("1"), therefore it must go first:
    * Solution so far: `(101, 1)`

  * The top now has an extra zero following, so tile 1 must go next since it's the only one having the bottom starting with a zero.
    * Solution so far: `(101, 1) (1, 0)`

  * The top now has an two extra ones following, so it's best for tile 3 to go next since it starts with two ones on the bottom.
    * Solution so far: `(101, 1) (1, 0) (0, 110)`

  * Now, the top and the bottom match up. The final solution is `(101, 1) (1, 0) (0, 110)` or tiles 2, 1, 3.

## 1(c, tool check): 

In [45]:
# Solution by tool for the same
pcp_solve([('1', '0'), ('101', '1'), ('0', '110')], OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[3, 1, 2]
0   1   101 
011 0   1   




The tool presented the solution in reverse, but we can see that it's the same solution.

## 1(d): 

In [57]:
# Create the tiles separately; easier to change and run that way
# Found it hard to create own instances that explode

TILES = [('101010101010', '10'), ('010101010101', '10'), ('01', '0101')]
pcp_solve(TILES, OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]
101010101010 010101010101 010101010101 010101010101 010101010101 010101010101 01           01           01           01           01           01           01           01           01           
10           10           10           10           10           10           0101         0101         0101         0101         0101         0101         0101         0101         0101         

01           01           01           01           01           01           01           01           01           01           01           01           01           01           01           
0101         0101         01

Solution length: 36 for all six solutions

# =============

# (2a) AR, 15%:

* Suppose there exists a decider $H$ for $\textit{Halt}_\text{TM}$.

  * Since $H$ is a decider, $H$ will always halt, and $H$ will accept $\left<M, w\right>$ if and only if $\left<M, w\right> \in \textit{Halt}_\text{TM}$, or in other words, $M$ halts on $w$.

* Create a program $D$ that takes input of a Turing machine $T$. Let $D$ halt and accept if $H(\left<T, T\right>)$ rejects, and let $D$ loop otherwise.

  * This means that $D$ halts if the Turing machine $T$ loops on input $T$ and loops if $T$ halts on $T$.

* Run the program $D$ with input of the $D$ itself. There are two possibilities of what $D$ would do:

  * $D$ halts and accepts, meaning that $H(\left<D, D\right>)$ rejected, but this rejection would mean that $D$ should not halt on input $D$, which creates a contradiction.

  * $D$ loops, meaning that $H(\left<D, D\right>)$ accepted, but this acceptance would mean that $D$ should halt and accept on input $D$, which creates a contradiction.

* Therefore, we obtain a contradiction under all cases.
  
* Thus, $\textit{Halt}_\text{TM}$ is undecidable. $\square$


# (2b) AR, 10%:

* The mapping reduction from $A_\text{TM}$ to $\textit{CFL}_\text{TM}$ is to emit this machine $M'$:

  * $M'$ takes input of a string $x$.

  * If $x$ is in the form of $0^n1^n2^n$, then $M'$ rejects $x$ immediately.

  * Otherwise, $M'$ runs $M$ on $w$:

    * If $M$ accepts $w$, then $M'$ accepts $x$.

    * If $M$ rejects $w$, then $M'$ rejects $x$.

* Assume that there exists a decider $D$ that decides whether $M'$ would accept some input $x$ or not:

  * If $D$ accepts, then $M'$ has a CFL, and $M$ must've accepted $w$.

  * If $D$ rejects, then $M'$ does not have a CFL, and $M$ must've rejected $w$.
  
* Thus, we have achieved our mapping reduction.

  * That is, $M'$ has a CFL if and only if $M$ accepts $w$.

# =============

# (3, XL, SV) 50%:  


## 3.a.i - describe $A, B, f$

* What is $A$ in this problem?

  * $A$ is the set of PCPs that have finite solutions.

* What is $B$ in this problem?

  * $B$ is the set of CFGs that have more than one parses for some strings.

* What is $f$ in this problem?

  * Given a PCP, $f$ is the process of constructing a CFG based on the contents of the tiles in the input PCP.

## 3.a.ii - Proof for the $x \in A \implies f(x) \in B$ part

<!-- * Proof steps - be thorough; it is worth 25 points! -->

* For any PCP $P$, $P \in A$ would mean that $P$ has a finite solution, and $f(P) \in B$ would mean that the CFG constructed by the process $f$ has more than one parses for some string.

* Let a PCP $P$ with tiles $\tau = \{(w_1, x_1), \cdots, (w_n, x_n)\}$ be given.

* Suppose that $P \in A$, meaning that $P$ has at least one finite solution. Let a solution be given that produces a string $s$.

* Since $s$ is a solution of $P$, $s$ can be made up from some of either $P$'s tile-tops or tile-bottoms. In other words, we can let $s = w_{j_1}\cdots w_{j_k} = x_{j_1} \cdots x_{j_k}$, where $1 \le j_1, \cdots, j_k \le n$.

* Let $a_1, \cdots, a_n$ be any symbols that have not appeared in any of the strings in $w_{1..n}$ and $x_{1..n}$.

* Let string $z = a_k \cdots a_1$, and let string $t = sz$.

  * In other words, $t = w_{j_1} \cdots w_{j_k}z = x_{j_1} \cdots x_{j_k}z$.

* This implies that the string $t$ has two parses in $f(P)$:

  * Parse 1:

  $$
    \begin{aligned}
      S &\implies S_A \\
        &\implies w_{j_1}S_Aa_1 \\
        &\implies w_{j_1}w_{j_2}S_Aa_2a_1 \\
        &\cdots \\
        &\implies w_{j_1} \cdots w_{(j_k-1)}S_Aa_{(k-1)} \cdots a_1 \\
        &\implies w_{j_1} \cdots w_{j_k}a_k \cdots a_1 = sz = t\\
    \end{aligned}
  $$

  * Parse 2:

  $$
    \begin{aligned}
      S &\implies S_B \\
        &\implies x_{j_1}S_Ba_1 \\
        &\implies x_{j_1}x_{j_2}S_Ba_2a_1 \\
        &\cdots \\
        &\implies x_{j_1} \cdots x_{(j_k-1)}S_Ba_{(k-1)} \cdots a_1 \\
        &\implies x_{j_1} \cdots x_{j_k}a_k \cdots a_1 = sz = t\\
    \end{aligned}
  $$

* Since $f(P)$ has two parses for some string, $f(P) \in B$.

* Therefore, $P \in A \implies f(P) \in B$. $\square$


## 3.b

<!-- * Proof steps - be thorough; it is worth 25 points! -->

* For any PCP $P$, $P \notin A$ would mean that $P$ does not have a finite solution, and $f(P) \notin B$ would mean that the CFG constructed by the process $f$ has at most one parse for any string.

* Let a PCP $P$ with tiles $\tau = \{(w_1, x_1), \cdots, (w_n, x_n)\}$ be given.

* Suppose that $P \notin A$, meaning that $P$ has no finite solution. This would mean that no string can be made up from either $P$'s tile-tops or tile-bottoms.

* In other words, given any string $s$:

  * If $s = w_{j_1}\cdots w_{j_k}$, where $1 \le j_1, \cdots, j_k \le n$, then $s \ne x_{j_1} \cdots x_{j_k}$.
  
  * Similarly, if $s = x_{l_1}\cdots x_{l_m}$, where $1 \le l_1, \cdots, l_m \le n$, then $s \ne w_{l_1} \cdots w_{l_m}$.

* Let $a_1, \cdots, a_n$ be any symbols that have not appeared in any of the strings in $w_{1..n}$ and $x_{1..n}$.

* Let string $z = a_k \cdots a_1$, and let string $t = sz$.

  * In other words, $t = w_{j_1} \cdots w_{j_k}z$, or $t = x_{l_1} \cdots x_{l_m}z$, or neither.

* This implies that the string $t$ has at most one parse in $f(P)$ because one of the following cases must be true:

  * Case 1, $s = w_{j_1} \cdots w_{j_k}$, then only one parse for $t$ is possible:

  $$
    \begin{aligned}
      S &\implies S_A \\
        &\implies w_{j_1}S_Aa_1 \\
        &\implies w_{j_1}w_{j_2}S_Aa_2a_1 \\
        &\cdots \\
        &\implies w_{j_1} \cdots w_{(j_k-1)}S_Aa_{(k-1)} \cdots a_1 \\
        &\implies w_{j_1} \cdots w_{j_k}a_k \cdots a_1 = sz = t\\
    \end{aligned}
  $$
  
    * The parse tree cannot go down $S \implies S_B$ because $t \ne x_{j_1} \cdots x_{j_k}a_k \cdots a_1$ (due to $s \ne x_{j_1} \cdots x_{j_k}$ as explained earlier) and the parse tree must be $k$ levels deep (due to $a_k \cdots a_1$ involve symbols that cannot otherwise be made from $x_{1..n}$).

  * Case 2, $s = x_{l_1} \cdots w_{l_m}$, then only one parse $t$ is possible:

  $$
    \begin{aligned}
      S &\implies S_B \\
        &\implies x_{l_1}S_Ba_1 \\
        &\implies x_{l_1}x_{l_2}S_Ba_2a_1 \\
        &\cdots \\
        &\implies x_{l_1} \cdots x_{(l_m-1)}S_Ba_{(m-1)} \cdots a_1 \\
        &\implies x_{l_1} \cdots x_{l_m}a_m \cdots a_1 = sz = t\\
    \end{aligned}
  $$
  
    * The parse tree cannot go down $S \implies S_A$ because $t \ne w_{l_1} \cdots w_{l_m}a_m \cdots a_1$ (due to $s \ne w_{l_1} \cdots w_{l_m}$ as explained earlier) and the parse tree must be $m$ levels deep (due to $a_m \cdots a_1$ involve symbols that cannot otherwise be made from $w_{1..n}$).

  * Case 3, neither of the above cases is true: no parse tree can exist because $t$ does not fit in either the form of $w_{q_1} \cdots w_{q_r}a_r \cdots a_1$ or $x_{q_1} \cdots x_{q_r}a_r \cdots a_1$ (which is required to be accepted by the CFG) as otherwise one of the above cases would be true.

* Since $f(P)$ has at most one parse for any string, $f(P) \notin B$.

* Therefore, $P \notin A \implies f(P) \notin B$. $\square$


# END
